File: 01/cad.cpp
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
double *linha_por_coluna(double **matriz, double *vetor);
double *coluna_por_coluna(double **matriz, double *vetor);
void vector_fill_random(double *array);
void matrix_fill_random(double **matriz);

int matriz_size;

int main(int argc, char *argv[])
{
    matriz_size = atoi(argv[1]);

    double **matriz = new double *[matriz_size];
    for (int i = 0; i < matriz_size; i++)
        matriz[i] = new double[matriz_size];

    double *vetor = new double[matriz_size];
    matrix_fill_random(matriz);
    vector_fill_random(vetor);

    /*
    printf("Matriz:\n");
    for (int i = 0; i < matriz_size; i++)
    {
        for (int j = 0; j < matriz_size; j++)
        {
            printf("%f ", matriz[i][j]);
        }
    }
    printf("\n");

    printf("Vector:\n");
    for (int i = 0; i < matriz_size; i++)
    {
        printf("%f ", vetor[i]);
    }
    printf("\n");
    */
    clock_t start, end;
    double cpu_time_used;

    start = clock();
    double *result = linha_por_coluna(matriz, vetor);
    end = clock();
    cpu_time_used = ((double)(end - start)) / CLOCKS_PER_SEC;
    printf("linha_por_coluna took %f seconds to execute \n", cpu_time_used);

    /*printf("Resultado:\n");
    for (int i = 0; i < matriz_size; i++)
    {
        printf("%f ", result[i]);
    }
    printf("\n");*/

    start = clock();
    double *result2 = coluna_por_coluna(matriz, vetor);
    end = clock();
    cpu_time_used = ((double)(end - start)) / CLOCKS_PER_SEC;
    printf("coluna_por_coluna took %f seconds to execute \n", cpu_time_used);

    /*printf("Resultado:\n");
    for (int i = 0; i < matriz_size; i++)
    {
        printf("%f ", result2[i]);
    }
    printf("\n");*/
}

double *linha_por_coluna(double **matriz, double *vetor)
{
    double *b = new double[matriz_size]{0};

    for (int i = 0; i < matriz_size; i++)
    {
        for (int j = 0; j < matriz_size; j++)
        {
            b[i] += matriz[i][j] * vetor[j];
        }
    }
    return b;
}

double *coluna_por_coluna(double **matriz, double *vetor)
{
    double *b = new double[matriz_size]{0};

    for (int i_v = 0; i_v < matriz_size; i_v++)
    {
        for (int i_m = 0; i_m < matriz_size; i_m++)
        {
            b[i_m] += matriz[i_m][i_v] * vetor[i_v];
        }
    }
    return b;
}

void vector_fill_random(double *array)
{
    for (int i = 0; i < matriz_size; i++)
    {
        double rng = 2 * ((double)rand() / (double)RAND_MAX) - 1;
        array[i] = rng;
    }
}

void matrix_fill_random(double **matriz)
{
    for (int i = 0; i < matriz_size; i++)
    {
        for (int j = 0; j < matriz_size; j++)
        {
            double rng = 2 * ((double)rand() / (double)RAND_MAX) - 1;
            matriz[i][j] = rng;
        }
    }
}

File: 01/cad.f90

program hello
    integer::matriz_size
    double precision, dimension(:,:), allocatable :: matriz 
    double precision, dimension(:), allocatable :: vetor
    double precision, dimension(:), allocatable :: result_lpc
    double precision, dimension(:), allocatable :: result_cpc
    real    ::  T1,T2 

    read *, matriz_size
    allocate(matriz(matriz_size,matriz_size))
    allocate(vetor(matriz_size))
    call matrix_fill_random(matriz)
    call vector_fill_random(vetor)

    !do z = 1, matriz_size
    !    do w = 1, matriz_size
    !        write(*,*) matriz(z,w)
    !    end do
    !end do
    !write(*,*) "\n"

    !do z = 1, matriz_size
    !    write(*,*) vetor(z)
    !end do
    !write(*,*) "\n"

    call cpu_time(T1)
    result_lpc = linha_por_coluna(matriz,vetor)
    call cpu_time(T2)
    print *, T2-T1

    !do z = 1, matriz_size
    !    write(*,*) result_lpc(z)
    !end do
    !write(*,*) "\n"

    call cpu_time(T1)
    result_cpc = coluna_por_coluna(matriz,vetor)
    call cpu_time(T2)
    print *, T2-T1

    !do z = 1, matriz_size
    !    write(*,*) result_cpc(z)
    !end do

contains

    subroutine vector_fill_random(array)
        double precision, dimension(:) :: array 
        double precision :: rng
        do n = 1, size(array)
            call random_number(rng)
            array(n) = rng
        end do
    end subroutine

    subroutine matrix_fill_random(matrix)
        double precision, dimension(:,:) :: matrix 
        double precision :: rng
        integer, dimension(2)  :: m_shape
        m_shape = shape(matrix)
        do n = 1, m_shape(1)
            do m = 1, m_shape(1)
                call random_number(rng)
                matrix(n,m) = rng
            end do
        end do
    end subroutine

    FUNCTION linha_por_coluna(matriz, vetor)  RESULT( b )

    double precision, dimension(:,:), intent(in) :: matriz
    double precision, dimension(:), intent(in) :: vetor 
    double precision, DIMENSION( SIZE(vetor) ) :: b
    b = 0.0

    do i = 1, size(vetor)
        do j = 1, size(vetor)
            b(i) = b(i) + matriz(i,j)*vetor(j)
        end do
    end do              
   END
   
   FUNCTION coluna_por_coluna(matriz, vetor)  RESULT( b )

    double precision, dimension(:,:), intent(in) :: matriz
    double precision, dimension(:), intent(in) :: vetor 
    double precision, DIMENSION( SIZE(vetor) ) :: b
    b = 0.0

    do i_v = 1, size(vetor)
        do i_m = 1, size(vetor)
            b(i_m) = b(i_m) + matriz(i_m,i_v)*vetor(i_v)
        end do
    end do              
   END

end program hello